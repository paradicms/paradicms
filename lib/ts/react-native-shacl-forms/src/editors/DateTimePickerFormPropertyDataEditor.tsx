import * as React from "react";
import DateTimePicker, {
  DateTimePickerEvent,
} from "@react-native-community/datetimepicker";
import {fromRdf, toRdf} from "rdf-literal";
import {Literal} from "@rdfjs/types";
import {useTheme} from "@rneui/themed";
import {View} from "react-native";
import {xsd} from "@paradicms/vocabularies";
import {DataFactory} from "@paradicms/rdf";
import {FormPropertyData} from "@paradicms/shacl-forms";

export const DateTimePickerFormPropertyDataEditor: React.FunctionComponent<{
  formPropertyData: FormPropertyData;
  mode: "date" | "time";
  onChange: () => void;
}> = ({formPropertyData, mode, onChange}) => {
  const value = formPropertyData.findAndMapValue(literal => {
    if (literal.termType === "Literal") {
      const value = fromRdf(literal as Literal);
      if (value instanceof Date) {
        return value;
      }
    }
    return null;
  });

  const {theme} = useTheme();

  return (
    <View
      style={{
        alignItems: "flex-start",
        marginTop: theme.spacing.md,
      }}
    >
      <DateTimePicker
        mode={mode}
        onChange={(event: DateTimePickerEvent, date: Date | undefined) => {
          if (date) {
            formPropertyData.value = toRdf(date, {
              dataFactory: DataFactory,
              datatype: mode === "date" ? xsd.date : xsd.time,
            });
          } else {
            formPropertyData.values = [];
          }
          onChange();
        }}
        style={{
          minWidth: 128,
        }}
        value={value ?? new Date()}
      />
    </View>
  );
};
