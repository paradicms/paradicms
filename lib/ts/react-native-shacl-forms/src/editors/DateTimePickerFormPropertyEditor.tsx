import * as React from "react";
import {FormProperty} from "@paradicms/shacl-forms";
import DateTimePicker, {
  DateTimePickerEvent,
} from "@react-native-community/datetimepicker";
import {fromRdf, toRdf} from "rdf-literal";
import {Literal} from "@rdfjs/types";
import {useTheme} from "@rneui/themed";
import {View} from "react-native";
import {xsd} from "@paradicms/vocabularies";
import {DataFactory} from "@paradicms/rdf";

export const DateTimePickerFormPropertyEditor: React.FunctionComponent<{
  formProperty: FormProperty;
  mode: "date" | "time";
  onChange: () => void;
}> = ({formProperty, mode, onChange}) => {
  const value = formProperty.findAndMapValue(literal => {
    if (literal.termType === "Literal") {
      const value = fromRdf(literal as Literal);
      if (value instanceof Date) {
        return value;
      }
    }
    return null;
  });

  const {theme} = useTheme();

  return (
    <View
      style={{
        alignItems: "flex-start",
        marginTop: theme.spacing.md,
      }}
    >
      <DateTimePicker
        mode={mode}
        onChange={(event: DateTimePickerEvent, date: Date | undefined) => {
          if (date) {
            formProperty.value = toRdf(date, {
              dataFactory: DataFactory,
              datatype: mode === "date" ? xsd.date : xsd.time,
            });
          } else {
            formProperty.values = [];
          }
          onChange();
        }}
        style={{
          minWidth: 128,
        }}
        value={value ?? new Date()}
      />
    </View>
  );
};
