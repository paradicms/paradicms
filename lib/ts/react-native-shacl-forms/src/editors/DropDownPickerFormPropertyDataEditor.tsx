import * as React from "react";
import {useState} from "react";
import {FormPropertyData} from "@paradicms/shacl-forms";
import {BlankNode, Literal, NamedNode} from "@rdfjs/types";
import {getRdfNodeLabel} from "@paradicms/rdf";
import DropDownPicker from "react-native-dropdown-picker";

export const DropDownPickerFormPropertyDataEditor: React.FunctionComponent<{
  formPropertyData: FormPropertyData;
  onChange: () => void;
}> = ({formPropertyData, onChange}) => {
  const [open, setOpen] = useState<boolean>(false);

  type Item = {
    label: string;
    value: string;
    valueTerm: BlankNode | Literal | NamedNode;
  };
  const items: Item[] = [];
  for (const valueTerm of formPropertyData.shape.in_ ?? []) {
    switch (valueTerm.termType) {
      case "BlankNode":
      case "NamedNode":
        items.push({
          label:
            getRdfNodeLabel({
              dataset: formPropertyData.dataGraph,
              node: valueTerm,
            }) ?? valueTerm.value,
          value: valueTerm.value,
          valueTerm,
        });
        break;
      case "Literal":
        items.push({
          label: valueTerm.value,
          value: valueTerm.value,
          valueTerm,
        });
      default:
        break;
    }
  }

  const value = formPropertyData.findAndMapValue(term =>
    items.some(option => option.valueTerm.equals(term)) ? term.value : null
  );

  return (
    <DropDownPicker
      dropDownDirection="TOP"
      items={items}
      open={open}
      setOpen={setOpen}
      onSelectItem={item => {
        formPropertyData.value = (item as Item).valueTerm;
        onChange();
      }}
      setValue={callback => {}}
      value={value}
    />
  );
};
