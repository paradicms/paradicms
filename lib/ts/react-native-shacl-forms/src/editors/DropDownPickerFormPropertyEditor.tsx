import * as React from "react";
import {useState} from "react";
import {FormProperty} from "@paradicms/shacl-forms";
import {BlankNode, Literal, NamedNode} from "@rdfjs/types";
import {getRdfNodeLabel} from "@paradicms/rdf";
import DropDownPicker from "react-native-dropdown-picker";

export const DropDownPickerFormPropertyEditor: React.FunctionComponent<{
  formProperty: FormProperty;
  onChange: () => void;
}> = ({formProperty, onChange}) => {
  const [open, setOpen] = useState<boolean>(false);

  const items: {
    label: string;
    value: string;
    valueTerm: BlankNode | Literal | NamedNode;
  }[] = [];
  for (const valueTerm of formProperty.shape.in_ ?? []) {
    switch (valueTerm.termType) {
      case "BlankNode":
      case "NamedNode":
        items.push({
          label:
            getRdfNodeLabel({
              dataset: formProperty.dataGraph,
              node: valueTerm,
            }) ?? valueTerm.value,
          value: valueTerm.value,
          valueTerm,
        });
        break;
      case "Literal":
        items.push({
          label: valueTerm.value,
          value: valueTerm.value,
          valueTerm,
        });
      default:
        break;
    }
  }

  const value = formProperty.findAndMapValue(term =>
    items.some(option => option.valueTerm.equals(term)) ? term.value : null
  );

  return (
    <DropDownPicker
      items={items}
      open={open}
      setOpen={setOpen}
      setValue={() => (value: string) => {
        for (const item of items) {
          if (value === item.value) {
            formProperty.value = item.valueTerm;
          }
        }
      }}
      value={value}
    />
  );
};
