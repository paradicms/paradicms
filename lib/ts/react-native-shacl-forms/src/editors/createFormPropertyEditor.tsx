import * as React from "react";
import {dash} from "@paradicms/vocabularies";
import {TextInputFormPropertyEditor} from "./TextInputFormPropertyEditor";
import {FormPropertyEditorFactory} from "./FormPropertyEditorFactory";

export const createFormPropertyEditor: FormPropertyEditorFactory = ({
  formProperty,
  onChange,
}) => {
  const editor = formProperty.editor;
  if (!editor) {
    console.warn(`form property ${formProperty.id} has no editor`);
    return null;
  }

  if (editor.equals(dash.TextFieldEditor)) {
    return (
      <TextInputFormPropertyEditor
        formProperty={formProperty}
        onChange={onChange}
      />
    );
  } else {
    console.warn(
      `form property ${formProperty.id} has unsupported editor ${editor.value}`
    );
    return null;
  }
};
