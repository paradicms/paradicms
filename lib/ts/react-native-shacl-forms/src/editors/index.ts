export * from "./createFormPropertyDataEditor";
export * from "./FormNodeDataEditor";
