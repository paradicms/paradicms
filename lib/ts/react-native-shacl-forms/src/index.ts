export * from "./editors";
export * from "./IconFactory";
export * from "./viewers";
