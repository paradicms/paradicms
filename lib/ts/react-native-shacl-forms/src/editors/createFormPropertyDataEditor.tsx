import * as React from "react";
import {dash} from "@paradicms/vocabularies";
import {TextInputFormPropertyDataEditor} from "./TextInputFormPropertyDataEditor";
import {FormPropertyDataEditorFactory} from "./FormPropertyDataEditorFactory";
import {DateTimePickerFormPropertyDataEditor} from "./DateTimePickerFormPropertyDataEditor";
import {DropDownPickerFormPropertyDataEditor} from "./DropDownPickerFormPropertyDataEditor";

export const createFormPropertyDataEditor: FormPropertyDataEditorFactory = ({
  formPropertyData,
  onChange,
}) => {
  const editor = formPropertyData.editor;
  if (!editor) {
    // console.warn(`form property ${formProperty.id} has no editor`);
    return null;
  }

  if (editor.equals(dash.DatePickerEditor)) {
    return (
      <DateTimePickerFormPropertyDataEditor
        formPropertyData={formPropertyData}
        mode="date"
        onChange={onChange}
      />
    );
  } else if (editor.equals(dash.EnumSelectEditor)) {
    return (
      <DropDownPickerFormPropertyDataEditor
        formPropertyData={formPropertyData}
        onChange={onChange}
      />
    );
  } else if (editor.equals(dash.TextFieldEditor)) {
    return (
      <TextInputFormPropertyDataEditor
        formPropertyData={formPropertyData}
        onChange={onChange}
      />
    );
  } else {
    console.warn(
      `form property ${formPropertyData.id} has unsupported editor ${editor.value}`
    );
    return null;
  }
};
