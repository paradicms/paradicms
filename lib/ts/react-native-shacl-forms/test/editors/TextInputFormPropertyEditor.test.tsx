import * as React from "react";
import * as renderer from "react-test-renderer";
import {testForm} from "../testForm";
import {schema} from "@paradicms/vocabularies";
import {TextInputFormPropertyDataEditor} from "../../src/editors";
import {FormProperty} from "@paradicms/shacl-forms";

test("renders correctly", () => {
  const formProperty = testForm().nodeTypes[0].nodes[0].properties.find(
    (property: FormProperty) => property.path.equals(schema.givenName)
  )!;
  const tree = renderer
    .create(
      <TextInputFormPropertyDataEditor
        formPropertyData={formProperty}
        onChange={() => {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
