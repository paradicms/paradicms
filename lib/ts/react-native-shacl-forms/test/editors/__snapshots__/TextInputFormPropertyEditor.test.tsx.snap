// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "paddingHorizontal": 10,
      "width": "100%",
    }
  }
  testID="RNE__Input__view-wrapper"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#86939e",
        "flexDirection": "row",
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <TextInput
      editable={true}
      onChangeText={[Function]}
      placeholder="given name"
      placeholderTextColor="#86939e"
      style={
        {
          "borderWidth": 1,
          "color": "#242424",
          "flex": 1,
          "fontSize": 18,
          "height": 40,
          "margin": 12,
          "minHeight": 40,
          "padding": 10,
        }
      }
      testID="RNE__Input__text-input"
      underlineColorAndroid="transparent"
      value="Minor"
    />
  </View>
  <Text
    style={
      {
        "color": "#ff190c",
        "fontSize": 12,
        "margin": 5,
      }
    }
  />
</View>
`;
